src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonHybridResumableConverter.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Text.Json.Serialization.Metadata;
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// Base class for converters that are able to resume after reading or writing to a buffer.
+    /// Writes initially start as non-resumable but are able to upgrade to resumable.
+    /// This is used when the Stream-based serialization APIs are used.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal abstract class JsonHybridResumableConverter<T> : JsonConverter<T>
```

Comment:
How is this different compared to the existing resumable converter base?

Reply:
It exposes `WriteWithoutStackFrame` which doesn't push/pop stack frames by default but based on the value (if it's long or serialization is done via streaming) there's a method to push/pop the stack frames. This keeps the common case of short strings/byte[] fast by not needing a stack frame but still allows large strings to opt into stack management.